<!-- [![Nextflow](https://img.shields.io/badge/nextflow%20DSL2-%E2%89%A524.04.2-23aa62.svg?labelColor=000000&logo=data:image/svg%2bxml;base64,PHN2ZyB3aWR0aD0iMjUxIiBoZWlnaHQ9IjI1MiIgdmlld0JveD0iMCAwIDI1MSAyNTIiIGZpbGw9Im5vbmUiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyI+DQo8cGF0aCBkPSJNMCA0Ny42MzQ1QzM5LjQ1IDUwLjI1NDMgNzEuMDYgODEuOTQyMiA3My41NCAxMjEuNDNIMTE5LjYxQzExNy4wNSA1Ni40NzM5IDY0LjkzIDQuMjU3NDQgMCAxLjU1NzYyVjQ3LjYzNDVaIiBmaWxsPSJ3aGl0ZSIvPg0KPHBhdGggZD0iTTczLjggMTMxLjkzOUM3MS4xOCAxNzEuMzg2IDM5LjQ5IDIwMi45OTQgMCAyMDUuNDc0VjI1MS41NDFDNjQuOTYgMjQ4Ljk4MSAxMTcuMTggMTk2Ljg2NSAxMTkuODggMTMxLjkzOUg3My44WiIgZmlsbD0id2hpdGUiLz4NCjxwYXRoIGQ9Ik0xNzYuMjAxIDEyMS4xNkMxNzguODIxIDgxLjcxMjIgMjEwLjUxMSA1MC4xMDQzIDI1MC4wMDEgNDcuNjI0NVYxLjU1NzYyQzE4NS4wNDEgNC4xMTc0NCAxMzIuODIxIDU2LjIzMzkgMTMwLjEyMSAxMjEuMTZIMTc2LjIwMVoiIGZpbGw9IndoaXRlIi8+DQo8cGF0aCBkPSJNMjUwLjAwMSAyMDUuNDY0QzIxMC41NTEgMjAyLjg0NSAxNzguOTQxIDE3MS4xNTcgMTc2LjQ2MSAxMzEuNjY5SDEzMC4zOTFDMTMyLjk1MSAxOTYuNjI1IDE4NS4wNzEgMjQ4Ljg0MiAyNTAuMDAxIDI1MS41NDFWMjA1LjQ2NFoiIGZpbGw9IndoaXRlIi8+DQo8L3N2Zz4=)](https://www.nextflow.io/) -->
[![Nextflow](https://img.shields.io/badge/nextflow%20DSL2-%E2%89%A521.04.1-23aa62.svg?labelColor=000000&logo=data:image/svg%2bxml;base64,PHN2ZyB3aWR0aD0iMjUxIiBoZWlnaHQ9IjI1MiIgdmlld0JveD0iMCAwIDI1MSAyNTIiIGZpbGw9Im5vbmUiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyI+DQo8cGF0aCBkPSJNMCA0Ny42MzQ1QzM5LjQ1IDUwLjI1NDMgNzEuMDYgODEuOTQyMiA3My41NCAxMjEuNDNIMTE5LjYxQzExNy4wNSA1Ni40NzM5IDY0LjkzIDQuMjU3NDQgMCAxLjU1NzYyVjQ3LjYzNDVaIiBmaWxsPSIjMjJBRTYzIi8+DQo8cGF0aCBkPSJNNzMuOCAxMzEuOTM5QzcxLjE4IDE3MS4zODYgMzkuNDkgMjAyLjk5NCAwIDIwNS40NzRWMjUxLjU0MUM2NC45NiAyNDguOTgxIDExNy4xOCAxOTYuODY1IDExOS44OCAxMzEuOTM5SDczLjhaIiBmaWxsPSIjMjJBRTYzIi8+DQo8cGF0aCBkPSJNMTc2LjIwMSAxMjEuMTZDMTc4LjgyMSA4MS43MTIyIDIxMC41MTEgNTAuMTA0MyAyNTAuMDAxIDQ3LjYyNDVWMS41NTc2MkMxODUuMDQxIDQuMTE3NDQgMTMyLjgyMSA1Ni4yMzM5IDEzMC4xMjEgMTIxLjE2SDE3Ni4yMDFaIiBmaWxsPSIjMjJBRTYzIi8+DQo8cGF0aCBkPSJNMjUwLjAwMSAyMDUuNDY0QzIxMC41NTEgMjAyLjg0NSAxNzguOTQxIDE3MS4xNTcgMTc2LjQ2MSAxMzEuNjY5SDEzMC4zOTFDMTMyLjk1MSAxOTYuNjI1IDE4NS4wNzEgMjQ4Ljg0MiAyNTAuMDAxIDI1MS41NDFWMjA1LjQ2NFoiIGZpbGw9IiMyMkFFNjMiLz4NCjwvc3ZnPg==)](https://www.nextflow.io/)
[![run with conda](https://img.shields.io/badge/run%20with-conda-3EB049.svg?labelColor=000000&logo=anaconda)](https://docs.conda.io/en/latest/)
[![run with docker](https://img.shields.io/badge/run%20with-docker-0db7ed.svg?labelColor=000000&logo=docker)](https://www.docker.com/)
[![run with apptainer/singularity](https://img.shields.io/badge/run%20with-apptainer%2Fsingularity-1E95D3.svg?labelColor=000000&logo=data:image/svg%2bxml;base64,PHN2ZyB3aWR0aD0iMjQ1IiBoZWlnaHQ9IjI0MCIgdmlld0JveD0iNjAgMCAzMTAgMjUwIiBmaWxsPSJub25lIiB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciPgk8cGF0aCBkPSJtIDI3MC4xOCwyNTMuOTggYyAtMS44LC0xLjIgLTMuNCwtMyAtNC40LC01LjIgbCAtNTIuNiwtMTE3LjQgYyAtMi4yLC00LjggLTMuOCwtOC42IC01LjIsLTExLjYgLTIuMiwtNC40IC0yLjIsLTUuNiAtMi4yLC02LjQgMCwtMi4yIDAuOCwtMy44IDIuNiwtNC44IHYgLTQuNCBoIC00My4yIHYgNC40IGMgMC44LDAuNCAxLjIsMS4yIDEuOCwxLjggMC40LDAuOCAwLjgsMS44IDAuOCwzIDAsMS4yIC0wLjQsMyAtMS44LDUuNiAtMS4yLDIuNiAtMi42LDUuNiAtNC40LDkuNCBsIC01MS44LDExNyBjIC0wLjgsMS44IC0yLjIsNC40IC0zLjgsNy40IC0xLjgsMyAtNC44LDQuNCAtOC4yLDQuOCB2IDMuOCBoIDQ5LjYgdiAtMy44IGMgLTUuNiwwIC04LjIsLTIuMiAtOC4yLC01LjYgMCwtMS44IDAuOCwtNC44IDMsLTkgMS44LC0zLjQgMy44LC03LjggNS42LC0xMiAyNC42LDkuNCA1Mi4yLDEwIDc2LjgsMC44IDIuMiw0LjQgMy44LDguMiA1LjIsMTEuMiAxLjgsMy40IDIuNiw2LjQgMi42LDguNiAwLDIuMiAtMC44LDMuOCAtMi4yLDQuOCAtMS4yLDAuNCAtMi4yLDAuOCAtMy40LDEuMiB2IDMuOCBoIDUwLjQgdiAtMy44IGMgLTIuOCwtMS44IC01LjQsLTIuOCAtNywtMy42IHogbSAtMTExLjQsLTQ3IDI3LjYsLTYxLjQgMjgsNjIuMiBjIC0xOCw2IC0zNy40LDYgLTU1LjYsLTAuOCB6IiBmaWxsPSJ3aGl0ZSIvPiA8cGF0aCBkPSJtIDg5Ljc4LDE0MC45OCBjIDAsLTkgMS4yLC0xNy42IDMuNCwtMjYuNCBsIC0yOCwtMTIuNiBjIC0zLjgsMTIgLTYsMjQuNiAtNiwzNy42IDAsMzUgMTQuMiw2OC42IDM5LjgsOTIuOCBsIDEuOCwtMy40IDExLjIsLTI1LjQgYyAtMTMuNiwtMTcuNCAtMjIuMiwtMzkgLTIyLjIsLTYyLjYgeiIgZmlsbD0iIzkzOTU5OCIvPiA8cGF0aCBkPSJtIDMxMC4xOCwxMDIuNTggLTI4LDEyLjYgYyAyLjIsOC4yIDMuNCwxNi44IDMuNCwyNS44IDAsMjMuOCAtOC42LDQ1LjggLTIyLjgsNjIuNiBsIDExLjYsMjUuNCAxLjgsMy40IGMgMjUuNCwtMjQuMiAzOS44LC01Ny44IDM5LjgsLTkyLjggLTAuMiwtMTIuNCAtMi4yLC0yNSAtNS44LC0zNyB6IiBmaWxsPSIjRjc5NDIxIi8+IDxwYXRoIGQ9Im0gNzEuMTgsODYuOTggMjcuNiwxMi42IGMgMTQuNiwtMzEgNDQuOCwtNTMgODAuMiwtNTYuMiB2IC0zMC42IGMgLTQ2LDIuNiAtODguNCwzMS40IC0xMDcuOCw3NC4yIHoiIGZpbGw9IiMxRTk1RDMiLz4gPHBhdGggZD0ibSAzMDQuMTgsODYuOTggYyAtMTkuNCwtNDIuOCAtNjEuOCwtNzEuNiAtMTA4LjQsLTc0LjYgdiAzMC42IGMgMzUuOCwzIDY2LDI1IDgwLjYsNTYuMiB6IiBmaWxsPSIjNkZCNTQ0Ii8+PC9zdmc+)](https://sylabs.io/docs/)
<!-- [![run with apptainer/singularity](https://img.shields.io/badge/run%20with-apptainer%2Fsingularity-1E95D3.svg?labelColor=000000&logo=data:image/svg%2bxml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9Ijg2LjQ5OTEgMTIgMTM3LjUgMTMzLjIiPg0KCTxwYXRoIGQ9Im0xOTkuNCAxNDEuNWMtMS0uNi0xLjgtMS41LTIuNC0yLjhsLTI4LjEtNjIuOWMtMS4xLTIuNS0yLjEtNC41LTIuOS02LjItMS4xLTIuMy0xLjItMy4xLTEuMi0zLjUgMC0xLjEuNS0yIDEuNC0yLjV2LTIuMmgtMjMuMXYyLjNjLjQuMy43LjYuOS45LjMuNC41LjkuNSAxLjYgMCAuNi0uMyAxLjYtMSAyLjktLjYgMS4zLTEuNCAzLTIuMyA1LjFsLTI3LjYgNjIuNmMtLjQgMS0xLjEgMi4yLTIuMSAzLjhzLTIuNSAyLjQtNC41IDIuNXYyLjFoMjYuNXYtMmMtMi45LS4xLTQuNC0xLjEtNC40LTMgMC0uOS42LTIuNSAxLjctNC44LjktMS45IDItNC4xIDMuMS02LjUgMTMuMSA1LjIgMjcuOSA1LjMgNDEuMS41IDEuMSAyLjMgMiA0LjMgMi44IDUuOS45IDEuOSAxLjQgMy40IDEuNCA0LjcgMCAxLjItLjQgMi4xLTEuMiAyLjUtLjYuMy0xLjIuNS0xLjkuNnYyLjFoMjcuMXYtMi4yYy0xLjYtLjUtMi45LTEtMy44LTEuNW0tNTkuOC0yNS4zIDE0LjctMzIuOCAxNS4xIDMzLjRjLTkuNSAzLjItMjAuMSAzLjEtMjkuOC0uNm0tMzYuOS0zNS40YzAtNC45LjctOS41IDEuOS0xNGwtMTQuOS02LjdjLTIuMSA2LjYtMy4xIDEzLjMtMy4yIDIwLS4xIDE4LjcgNy43IDM2LjggMjEuMyA0OS43bDEtMS45IDYuMS0xMy43Yy03LjYtOS0xMi4yLTIwLjctMTIuMi0zMy40bTExOC4xLTIwLjYtMTQuOSA2LjdjMS4yIDQuNCAxLjkgOS4xIDEuOSAxMy45IDAgMTIuOC00LjYgMjQuNS0xMi4yIDMzLjZsNi4xIDEzLjcgMSAxLjhjMTMuNi0xMi45IDIxLjQtMzEgMjEuMy00OS43LS4xLTYuOC0xLjItMTMuNC0zLjItMjBtLTEyNy45LTguMyAxNC44IDYuNmM3LjgtMTYuNiAyNC0yOC40IDQzLjEtMzB2LTE2LjVjLTI1IDEuNS00Ny42IDE3LjEtNTcuOSAzOS45bTEyNC43LjFjLTEwLjQtMjIuOC0zMy0zOC40LTU4LTQwdjE2LjVjMTkuMiAxLjYgMzUuNCAxMy40IDQzLjEgMzAuMWwxNC45LTYuNiIgZmlsbD0id2hpdGUiLz4NCjwvc3ZnPg==)](https://sylabs.io/docs/) -->


<!-- Nextflow logo -->
<!-- <svg width="251" height="252" viewBox="0 0 251 252" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M0 47.6345C39.45 50.2543 71.06 81.9422 73.54 121.43H119.61C117.05 56.4739 64.93 4.25744 0 1.55762V47.6345Z" fill="#22AE63"/>
<path d="M73.8 131.939C71.18 171.386 39.49 202.994 0 205.474V251.541C64.96 248.981 117.18 196.865 119.88 131.939H73.8Z" fill="#22AE63"/>
<path d="M176.201 121.16C178.821 81.7122 210.511 50.1043 250.001 47.6245V1.55762C185.041 4.11744 132.821 56.2339 130.121 121.16H176.201Z" fill="#22AE63"/>
<path d="M250.001 205.464C210.551 202.845 178.941 171.157 176.461 131.669H130.391C132.951 196.625 185.071 248.842 250.001 251.541V205.464Z" fill="#22AE63"/>
</svg> -->

<!-- Apptainer Medium version -->
<!-- 
<svg width="245" height="240" viewBox="60 0 310 250" fill="none" xmlns="http://www.w3.org/2000/svg">	<path d="m 270.18,253.98 c -1.8,-1.2 -3.4,-3 -4.4,-5.2 l -52.6,-117.4 c -2.2,-4.8 -3.8,-8.6 -5.2,-11.6 -2.2,-4.4 -2.2,-5.6 -2.2,-6.4 0,-2.2 0.8,-3.8 2.6,-4.8 v -4.4 h -43.2 v 4.4 c 0.8,0.4 1.2,1.2 1.8,1.8 0.4,0.8 0.8,1.8 0.8,3 0,1.2 -0.4,3 -1.8,5.6 -1.2,2.6 -2.6,5.6 -4.4,9.4 l -51.8,117 c -0.8,1.8 -2.2,4.4 -3.8,7.4 -1.8,3 -4.8,4.4 -8.2,4.8 v 3.8 h 49.6 v -3.8 c -5.6,0 -8.2,-2.2 -8.2,-5.6 0,-1.8 0.8,-4.8 3,-9 1.8,-3.4 3.8,-7.8 5.6,-12 24.6,9.4 52.2,10 76.8,0.8 2.2,4.4 3.8,8.2 5.2,11.2 1.8,3.4 2.6,6.4 2.6,8.6 0,2.2 -0.8,3.8 -2.2,4.8 -1.2,0.4 -2.2,0.8 -3.4,1.2 v 3.8 h 50.4 v -3.8 c -2.8,-1.8 -5.4,-2.8 -7,-3.6 z m -111.4,-47 27.6,-61.4 28,62.2 c -18,6 -37.4,6 -55.6,-0.8 z" fill="white"/> <path d="m 89.78,140.98 c 0,-9 1.2,-17.6 3.4,-26.4 l -28,-12.6 c -3.8,12 -6,24.6 -6,37.6 0,35 14.2,68.6 39.8,92.8 l 1.8,-3.4 11.2,-25.4 c -13.6,-17.4 -22.2,-39 -22.2,-62.6 z" fill="#939598"/> <path d="m 310.18,102.58 -28,12.6 c 2.2,8.2 3.4,16.8 3.4,25.8 0,23.8 -8.6,45.8 -22.8,62.6 l 11.6,25.4 1.8,3.4 c 25.4,-24.2 39.8,-57.8 39.8,-92.8 -0.2,-12.4 -2.2,-25 -5.8,-37 z" fill="#F79421"/> <path d="m 71.18,86.98 27.6,12.6 c 14.6,-31 44.8,-53 80.2,-56.2 v -30.6 c -46,2.6 -88.4,31.4 -107.8,74.2 z" fill="#1E95D3"/> <path d="m 304.18,86.98 c -19.4,-42.8 -61.8,-71.6 -108.4,-74.6 v 30.6 c 35.8,3 66,25 80.6,56.2 z" fill="#6FB544"/></svg>  -->



## Introduction

**nf-taxblast** is a [Nextflow](https://www.nextflow.io/) pipeline that uses a "split-process-combine" approach to efficiently perform homology searches of large multi-sequence fasta files (entire transcriptome or proteome) using DIAMOND<sup>1</sup> or NCBI BLAST<sup>23</sup>. This works best on high performance computing (HPC) clusters, which can utilise multiple compute nodes to process each "chunk" of query sequences in parallel. The workflow exposes all (or at least most) of the command-line options of either tool to the user, allowing full control of the output (including custom output formats and using the "tax-aware" v5 NCBI databases to report taxonmic classifications of the returned hits). 


## Installation
Clone this repository to your system with `git clone https://github.com/IdoBar/nf-taxblast.git` (don't forget where you downloaded it to!)

> [!NOTE]  
> If you are new to Nextflow<sup>4</sup> and nf-core<sup>5</sup>, please refer to [this page](https://nf-co.re/docs/usage/installation) on how to set-up Nextflow. Make sure to [test your setup](https://nf-co.re/docs/usage/introduction#how-to-run-a-pipeline) with `-profile test` before running the workflow on actual data. You will likely need to adapt one the [institutional profiles](https://nf-co.re/configs) provided by `nf-core` to your specific HPC, read more about it [here](https://nf-core-docs.netlify.app/docs/tutorials/use_nf-core_pipelines/writing_institutional_profiles) and contact your HPC admin if you need further assistance (I can probably help if you use any of the Australian HPC systems). 

## Usage

### Required input:

- Nucleotide or protein multi-sequence file in a `fasta` format.  
- A local repository of NCBI BLAST databases (v5), use the `--download` flag to download the required databases before running the homology search.

Once you have the databases, you can run the pipeline using:

```bash
nextflow run nf-taxblast.nf \
   --app blastn --query QUERY.fasta \
   --chunkSize 200 --db "path-of-db/db" \
   --out QUERY.blastn.db.outfmt6 \
   --outDir <OUTDIR> \
   -profile conda,blastn_tax \
```

To see the complete usage information, run `nextflow run <path-of-nf-taxblast>/nf-taxblast.nf --help`, which will print the following usage information:

```
   Usage:
      The typical command for running the pipeline is as follows:
      nextflow run nf-taxblast.nf --app blastn --query QUERY.fasta --chunkSize 200 --db "path-of-db/db" -profile conda,blastn_tax
      nextflow run nf-taxblast.nf --app "diamond blastp" --query QUERY.faa --chunkSize 5000 --db "path-of-db/db" -profile docker,diamond_tax

      Mandatory arguments:
       --app <value>                  BLAST/DIAMOND program to use (diamond blastp/x must be quoted!)
                                      Valid options: [blastn, blastp, tblastn, blastx, 'diamond blastp', 'diamond blastx'] 
       --query <file.fatsa>           Query fasta file of sequences you wish to BLAST
       --db <path-of-db/db>           Path of the BLAST or DIAMOND database. 
                                      If BLAST database is provided for DIAMOND and taxonomy information is requested
                                      then a suitable database will be created (see Taxonomy options below). 
                                      Default: [$BLASTDB/nt or $BLASTDB/nr for protein search]   
       -profile <profile1,profile2>   Configuration profile to use. Can use multiple (comma separated)
                                      Available profiles for container systems: [conda/apptainer/singularity/local/docker]
                                      Available profiles with preset database and output formats: [blastn_tax/diamond_tax]
                                      Available profiles with test datasets and databases: [test/test_tax/test_p/test_p_tax/test_d/test_d_tax]

       Optional arguments:
       --out <outfile.outfmt6>        Output filename of final BLAST output. Default: [QUERY.app.db.outfmt6]
       --outDir <path>                Output folder for the results. Default: [results]
       --outCols <'std'>              Output columns (must be quoted!). Default: ['std']
       --headers <false>              Include headers in the output table. Default: false
       --blastOpts <'-evalue 10'>     Additional options for BLAST command (must be quoted!). 
                                      Default: ['-evalue 1e-10 -max_target_seqs 20']
       --dmndOpts <'-e 10e-10'>       Additional options for BLAST command (must be quoted!). 
                                      Default: ['-e 1e-10 -k 20'] 
       --chunkSize <num>              Number of fasta records to use in each job when splitting the query fasta file. 
                                      This option can also take the size of each subquery (like 200.KB, 5.KB, etc.) 
                                      Default: [250]
       --queueSize <num>              Maximum number of jobs to be queued [50]
       --download <false>             Download database before running homology search. Default: false

       Taxonomy options:
       --taxDbDir <path-of-taxdb/db>  Location of taxonomy db files (prot.accession2taxid.FULL.gz, nodes.dmp and names.dmp) 
                                      to allow DIAMOND return taxonomic information columns. 
                                      If the required files cannot be found in the path they will be automatically downloaded 
                                      from the NCBI.
                                      Information about the required files and where to download them can be found at 
                                      https://github.com/bbuchfink/diamond/wiki/3.-Command-line-options#makedb-options
                                      Default: [same path as the database]
       --taxListFile <taxid.list>     A file with list of taxonomy IDs to limit the search space.
       --help                         This usage statement.
     
```



> [!WARNING]  
> Please provide pipeline parameters via the CLI or Nextflow `-params-file` option. Custom config files including those provided by the `-c` Nextflow option can be used to provide any configuration _**except for parameters**_; see the [nf-core docs](https://nf-co.re/docs/usage/getting_started/configuration#custom-configuration-files).

### Pipeline output
By default, the pipeline generates a tab-delimited file, with the combined top 20 hits (homologs) found for each of the query sequences (with a threshold E-value of 1e-10). This output format is commony known as `outfmt 6 std`, and includes the standard 12 columns (`qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore`, see [BLASTn output format 6](https://www.metagenomics.wiki/tools/blast/blastn-output-format-6)). Other columns can be added with the `--outCols` option, according to the specifications in the [BLAST documentation](https://www.ncbi.nlm.nih.gov/books/NBK279684/table/appendices.T.options_common_to_all_blast/) and [DIAMOND documentation](https://github.com/bbuchfink/diamond/wiki/3.-Command-line-options#output-options). Personally, I prefer to always include the `stitle` column, which is the description of the hit sequence.     

> [!NOTE]  
> Please note that the default results table **doesn't include any headers**. You can add the headers by using the `--headers` flag (which will save the output with the `.outfmt7` suffix). 
<!-- TODO - Add option to include headers -->


### Taxonomy options
#### Adding taxonomy information columns
One of the unique features of `nf-taxblast` (compared with other Nextflow implementations or BLAST) is that it allows the user to add columns with taxonomic classifications of the results by using the `--outCols` option. Some options include `staxids` (Subject Taxonomy ID(s), separated by a ';'), `sscinames` (Subject Scientific Name(s), separated by a ';'), `scomnames` (Subject Common Name(s), separated by a ';'), and `sskingdoms` (Subject Super Kingdom(s), separated by a ';'). These can be included in the results' table by using the `(blastn/blastp/diamond)_tax` profiles.

DIAMOND also offers to add taxonomy columns similar to BLAST, with slight differences: `skingdoms` (Unique Subject Kingdom(s), separated by a ';', not to be confused with `sskingdoms`) and `sphylums` (Unique Subject Phylums(s), separated by a ';') are available *in addition* to the BLAST taxonomy columns mentioned above, however **`scomnames` is not available**. 

DIAMOND, however, can not use the default NCBI databases to include taxonomic features and must create its own version of the database. If taxonomy columns are requested, `nf-taxblast` will look for a file named `db.dmnd` in the DB folder and if it can't find it, it will download the NCBI [prot.accession2taxid.FULL.gz](https://ftp.ncbi.nlm.nih.gov/pub/taxonomy/accession2taxid/prot.accession2taxid.FULL.gz) mapping file, extract the sequences from the requested database in `FASTA` format and will create the required `db.dmnd` file, as detailed in [DIAMOND makedb-options](https://github.com/bbuchfink/diamond/wiki/3.-Command-line-options#makedb-options).  
DIAMOND can also be used just for taxonomic classification by using `--outCols 102`. This will print only the Query ID (`qseqid`), NCBI taxid (`staxids`) and the E-value (`evalue`) of the best alignment (without the matching sequences' ID or any other columns). A fourth column containing the taxonomic lineage in text form can be added by using the option `--dmndOpts '--include-lineage'`.

<!-- > [!NOTE]  
> Please note that due to recent changes on the NCBI [prot.accession2taxid.FULL.gz](https://ftp.ncbi.nlm.nih.gov/pub/taxonomy/accession2taxid/prot.accession2taxid.FULL.gz) mapping file, only the most latest version of DIAMOND ([v2.1.12](https://github.com/bbuchfink/diamond/releases/tag/v2.1.12)) can parse it and make the custom DB and this version is not yet available in a Docker/Singularity/Apptainer container, so please use `-profile conda`.  -->

#### Limiting the search to specific taxonomic lineages
Another useful feature allows the user to limit the database to specific taxonomic lineages. This is useful when the user seeks to find hits from certain taxonomic groups (see more advantages and details in [this article](https://sequenceserver.com/blog/blast-with-taxids/)). To use this feature, use the `--taxListFile taxids.list` option to provide a text file with a list of taxonomy ids (each in a separate row, can be at any taxonomic level, from phylum to species). You can use the NCBI Taxonomy [Tax Identifier](https://www.ncbi.nlm.nih.gov/Taxonomy/TaxIdentifier/tax_identifier.cgi) or other tools, such as the [ETE Toolkit](https://etetoolkit.org/docs/latest/tutorial/tutorial_ncbitaxonomy.html) (python) or [taxize](https://docs.ropensci.org/taxize/) (R) to find the `taxid` for your species/genus/family, etc.


### Examples

Download the NCBI mRNA Refseq database and run `blastn` against it, generating an output table with header and including taxonomy information columns.
```bash
nextflow run nf-taxblast.nf \
   --app blastn --query QUERY.fasta \
   --chunkSize 200 --db "refseq_rna" --download \
   --out QUERY.blastn.refseq_rna.outfmt7 --headers \
   --outDir example \
   -profile conda,blastn_tax \
```


#### Running tests
Test the pipeline with `blastn`
```bash
nextflow run nf-taxblast.nf -profile conda,test
```

Test the pipeline with `blastn` and additional taxonomy columns
```bash
nextflow run nf-taxblast.nf -profile conda,test_tax
```

Test the pipeline with `blastp` and singularity
```bash
nextflow run nf-taxblast.nf -profile singularity,test_p
```

Test the pipeline with `blastp`, apptainer and additional taxonomy columns
```bash
nextflow run nf-taxblast.nf -profile apptainer,test_p_tax
```
Test the pipeline with `diamond` and apptainer
```bash
nextflow run nf-taxblast.nf -profile apptainer,test_d
```

Test the pipeline with `diamond` and additional taxonomy columns
```bash
nextflow run nf-taxblast.nf -profile conda,test_d_tax
```


### Monitor progress and utilised resources
Nextflow provides an easy method to monitor the progress of the pipeline, status of the tasks and resource usage via [Seqera Tower](https://cloud.seqera.io). This can be enabled with the `-with-tower` flag.  Please read the instructions on how to create a Tower API key and use it to monitor the runs [here](https://training.nextflow.io/2.1/basic_training/seqera_platform/). 


## Credits

`nf-taxblast` was inspired by an earlier work I did ([blast_tax](https://gist.github.com/IdoBar/5db8973e99463e4c41a10564780cfe6d)), implemented in Nextflow following the example workflow demonstrated in https://github.com/nextflow-io/blast-example, with improvements to expose more options to the user and allow taxonomic classification of the results.

## Contributions and Support

If you would like to contribute to this pipeline, please contact me, [raise an Issue](https://github.com/IdoBar/nf-taxblast/issues) or fork this repo, edit it and suggest a PR.

## Citations

Tools used in the pipeline: 

1. Buchfink, B., Xie, C. & Huson, D. H. Fast and sensitive protein alignment using DIAMOND. Nat Meth 12, 59–60 (2015). [10.1038/nmeth.3176](https://doi.org/10.1038/nmeth.3176)  
2. Camacho, C. et al. BLAST+: architecture and applications. BMC Bioinformatics 10, 421 (2009). [10.1186/1471-2105-10-421](https://doi.org/10.1186/1471-2105-10-421)  
3. Camacho, C. et al. BLAST® Command Line Applications User Manual. (National Center for Biotechnology Information (US), Bethesda, MD, USA, 2013). [link](http://www.ncbi.nlm.nih.gov/books/NBK1763/)  


Nextflow publications:

4. Langer, B. E. et al. Empowering bioinformatics communities with Nextflow and nf-core. bioRxiv Preprint (2024). [10.1101/2024.05.10.592912](https://doi.org/10.1101/2024.05.10.592912) 
5. Di Tommaso, P. et al. Nextflow enables reproducible computational workflows. Nature Biotechnology 35, 316–319 (2017). [10.1038/nbt.3820](https://doi.org/10.1038/nbt.3820)

Additional reading on similar alternatives and performance comparison:

6. Yim, W. C. & Cushman, J. C. Divide and Conquer (DC) BLAST: fast and easy BLAST execution within HPC environments. PeerJ 5, (2017).[10.7717/peerj.3486](https://doi.org/10.7717/peerj.3486)
7. Hernández-Salmerón, J. E. & Moreno-Hagelsieb, G. Progress in quickly finding orthologs as reciprocal best hits: comparing blast, last, diamond and MMseqs2. BMC Genomics 21, 741 (2020). [10.1186/s12864-020-07132-6](https://doi.org/10.1186/s12864-020-07132-6) 



<!-- > **The nf-core framework for community-curated bioinformatics pipelines.**
>
> Philip Ewels, Alexander Peltzer, Sven Fillinger, Harshil Patel, Johannes Alneberg, Andreas Wilm, Maxime Ulysse Garcia, Paolo Di Tommaso & Sven Nahnsen.
>
> _Nat Biotechnol._ 2020 Feb 13. doi: [10.1038/s41587-020-0439-x](https://dx.doi.org/10.1038/s41587-020-0439-x). -->